File: internal/hooks/stop.go
// stop hook handler.
// Ports runStop from scripts/hooks/stop.js.
// Always exits 0 and emits {"continue":true}. Uses internal/store + internal/yamllite.
package hooks

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/mretamozo-hiberuscom/ospec-workflow/internal/store"
	"github.com/mretamozo-hiberuscom/ospec-workflow/internal/yamllite"
)

func init() {
	Register(&stopHandler{})
}

type stopHandler struct{}

func (h *stopHandler) Name() string { return "stop" }

// stopInput is the stdin payload for stop.
type stopInput struct {
	Cwd       string `json:"cwd"`
	Timestamp string `json:"timestamp"`
	SessionID string `json:"sessionId"`
	SessionIDUnderscore string `json:"session_id"`
}

type stopOutput struct {
	Continue     bool    `json:"continue"`
	Status       string  `json:"status,omitempty"`
	Path         string  `json:"path,omitempty"`
	ActiveChange *string `json:"activeChange,omitempty"`
	SystemMessage string `json:"systemMessage,omitempty"`
}

func (h *stopHandler) Run(stdin []byte) ([]byte, int) {
	var input stopInput
	if err := json.Unmarshal(stdin, &input); err != nil {
		return continueWithError(fmt.Sprintf("Stop hook could not write the session trace: %s", err.Error())), 0
	}

	if err := runStop(input); err != nil {
		return continueWithError(fmt.Sprintf("Stop hook could not write the session trace: %s", err.Error())), 0
	}

	b, _ := json.Marshal(map[string]bool{"continue": true})
	return b, 0
}

func runStop(input stopInput) error {
	workspace := resolveCwd(input.Cwd)
	s := store.NewStore(workspace)

	changes, err := s.FindActiveChanges()
	if err != nil {
		return err
	}

	var activeChange *store.ActiveChange
	if len(changes) > 0 {
		activeChange = changes[0]
	}

	changeName := ""
	currentPhase := ""
	status := ""
	nextRecommended := ""

	if activeChange != nil {
		changeName = yamllite.ExtractFirstScalar(activeChange.Content, [][]string{
			{"change", "name"},
		})
		if changeName == "" {
			changeName = activeChange.DirectoryName
		}
		currentPhase = yamllite.ExtractFirstScalar(activeChange.Content, [][]string{
			{"change", "current_phase"},
			{"current_phase"},
			{"phase"},
		})
		status = yamllite.ExtractFirstScalar(activeChange.Content, [][]string{
			{"change", "status"},
			{"status"},
		})
		nextRecommended = yamllite.ExtractFirstScalar(activeChange.Content, [][]string{
			{"next_recommended"},
		})
	}

	// Resolve timestamp.
	ts := strings.TrimSpace(input.Timestamp)
	if ts == "" {
		ts = time.Now().UTC().Format(time.RFC3339Nano)
	}

	// Resolve session ID — camelCase takes priority, then underscore.
	sessionID := strings.TrimSpace(input.SessionID)
	if sessionID == "" {
		sessionID = strings.TrimSpace(input.SessionIDUnderscore)
	}
	if sessionID == "" {
		sessionID = "unknown"
	}

	// Detailed summary path.
	detailedSummary := "None"
	if activeChange != nil {
		summaryPath := s.SessionSummaryPath(activeChange.DirectoryName)
		if info, err := os.Stat(summaryPath); err == nil && info.Mode().IsRegular() {
			rel, err := filepath.Rel(workspace, summaryPath)
			if err == nil {
				detailedSummary = filepath.ToSlash(rel)
			} else {
				detailedSummary = filepath.ToSlash(summaryPath)
			}
		}
	}

	// Resolve next action.
	var nextAction string
	if activeChange != nil {
		nextAction = yamllite.FormatNextAction(nextRecommended, changeName)
	} else {
		nextAction = "Start a new session when more work is needed."
	}

	latestContent := renderLatestSummary(renderLatestArgs{
		hasChange:    activeChange != nil,
		changeName:   changeName,
		currentPhase: currentPhase,
		status:       status,
		detailedSummary: detailedSummary,
		endedAt:      ts,
		sessionID:    sessionID,
		nextAction:   nextAction,
	})

	latestPath := s.LatestSessionPath()
	if err := os.MkdirAll(filepath.Dir(latestPath), 0755); err != nil {
		return fmt.Errorf("stop: mkdir: %w", err)
	}
	return os.WriteFile(latestPath, []byte(latestContent), 0644)
}

// renderLatestArgs bundles all rendering parameters.
type renderLatestArgs struct {
	hasChange       bool
	changeName      string
	currentPhase    string
	status          string
	detailedSummary string
	endedAt         string
	sessionID       string
	nextAction      string
}

// renderLatestSummary ports renderLatestSummary from stop.js.
func renderLatestSummary(a renderLatestArgs) string {
	activeChangeVal := "`None`"
	if a.hasChange {
		activeChangeVal = "`" + a.changeName + "`"
	}
	currentPhaseVal := "`None`"
	if a.hasChange {
		p := a.currentPhase
		if p == "" {
			p = "unknown"
		}
		currentPhaseVal = "`" + p + "`"
	}
	statusVal := "`None`"
	if a.hasChange {
		st := a.status
		if st == "" {
			st = "active"
		}
		statusVal = "`" + st + "`"
	}

	lines := []string{
		"# Latest Session",
		"",
		"- Ended at: `" + a.endedAt + "`",
		"- Session: `" + a.sessionID + "`",
		"- Active change: " + activeChangeVal,
		"- Current phase: " + currentPhaseVal,
		"- Change status: " + statusVal,
		"- Detailed summary: `" + a.detailedSummary + "`",
		"",
		"## Next recommended action",
		a.nextAction,
		"",
	}
	return strings.Join(lines, "\n")
}

File: plugins/ospec-workflow/scripts/hooks/stop.js
#!/usr/bin/env node

"use strict";

const fs = require("node:fs/promises");
const path = require("node:path");

const {
  extractFirstScalar,
  formatNextAction,
} = require("./pre-compact.js");
const {
  ARTIFACT_STORE_RELATIVE_PATHS,
  createArtifactStoreFromConfig,
} = require("../lib/artifact-store.js");
const { writeFileAtomic } = require("../lib/atomic-write.js");
const { resolveWorkspaceCwd } = require("../lib/pathsafe.js");

const LATEST_RELATIVE_PATH = ARTIFACT_STORE_RELATIVE_PATHS.latestSession;

function toPortablePath(filePath) {
  return filePath.split(path.sep).join("/");
}

async function pathIsFile(filePath) {
  try {
    return (await fs.stat(filePath)).isFile();
  } catch (error) {
    if (error.code === "ENOENT") {
      return false;
    }

    throw error;
  }
}

function resolveTimestamp(input, now) {
  const timestamp =
    typeof input?.timestamp === "string" ? input.timestamp.trim() : "";

  return timestamp || now().toISOString();
}

function resolveSessionId(input) {
  const sessionId =
    typeof input?.sessionId === "string"
      ? input.sessionId.trim()
      : typeof input?.session_id === "string"
        ? input.session_id.trim()
        : "";

  return sessionId || "unknown";
}

async function resolveDetailedSummary(store, activeChange) {
  if (!activeChange) {
    return "None";
  }

  const summaryPath = store.sessionSummaryPath(activeChange.directoryName);

  return (await pathIsFile(summaryPath))
    ? toPortablePath(path.relative(store.workspace, summaryPath))
    : "None";
}

function renderLatestSummary({
  activeChange,
  changeName,
  currentPhase,
  detailedSummary,
  endedAt,
  nextAction,
  sessionId,
  status,
}) {
  return [
    "# Latest Session",
    "",
    `- Ended at: \`${endedAt}\``,
    `- Session: \`${sessionId}\``,
    `- Active change: \`${activeChange ? changeName : "None"}\``,
    `- Current phase: \`${activeChange ? currentPhase || "unknown" : "None"}\``,
    `- Change status: \`${activeChange ? status || "active" : "None"}\``,
    `- Detailed summary: \`${detailedSummary}\``,
    "",
    "## Next recommended action",
    nextAction,
    "",
  ].join("\n");
}

async function writeLatestTrace(filePath, content) {
  // Atomic temp+rename so a crash mid-write never leaves the next session
  // reading a partially written latest.md.
  await writeFileAtomic(filePath, content);
}

async function runStop({
  input = {},
  fallbackCwd = process.cwd(),
  mode,
  now = () => new Date(),
} = {}) {
  const workspace = resolveWorkspaceCwd(input.cwd, fallbackCwd);
  const store = await createArtifactStoreFromConfig({ mode, workspace });
  const activeChange = (await store.findActiveChanges())[0] || null;
  const changeName = activeChange
    ? extractFirstScalar(activeChange.content, [["change", "name"]]) ||
      activeChange.directoryName
    : "";
  const currentPhase = activeChange
    ? extractFirstScalar(activeChange.content, [
        ["change", "current_phase"],
        ["current_phase"],
        ["phase"],
      ])
    : "";
  const status = activeChange
    ? extractFirstScalar(activeChange.content, [
        ["change", "status"],
        ["status"],
      ])
    : "";
  const nextRecommended = activeChange
    ? extractFirstScalar(activeChange.content, [["next_recommended"]])
    : "";
  const latestPath = store.latestSessionPath();
  const latestSummary = renderLatestSummary({
    activeChange,
    changeName,
    currentPhase,
    detailedSummary: await resolveDetailedSummary(store, activeChange),
    endedAt: resolveTimestamp(input, now),
    nextAction: activeChange
      ? formatNextAction(nextRecommended, changeName)
      : "Start a new session when more work is needed.",
    sessionId: resolveSessionId(input),
    status,
  });

  await writeLatestTrace(latestPath, latestSummary);

  return {
    status: "written",
    path: LATEST_RELATIVE_PATH,
    activeChange: changeName || null,
  };
}

async function readJsonInput(stream = process.stdin) {
  const chunks = [];

  for await (const chunk of stream) {
    chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
  }

  const input = Buffer.concat(chunks).toString("utf8").trim();
  return input ? JSON.parse(input) : {};
}

async function main() {
  try {
    await runStop({ input: await readJsonInput() });
    process.stdout.write('{"continue":true}\n');
  } catch (error) {
    process.stdout.write(
      `${JSON.stringify({
        continue: true,
        systemMessage: `Stop hook could not write the session trace: ${error.message}`,
      })}\n`,
    );
  }
}

if (require.main === module) {
  void main();
}

module.exports = {
  LATEST_RELATIVE_PATH,
  renderLatestSummary,
  runStop,
};
